src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -611,13 +611,34 @@ public void AppendConflictingField(EnumFieldInfo other)
                 // Walk the existing list to ensure we do not add duplicates.
                 foreach (EnumFieldInfo conflictingField in conflictingFields)
                 {
-                    if (conflictingField.Kind is EnumFieldNameKind.Default || conflictingField.JsonName.Equals(other.JsonName, StringComparison.Ordinal))
+                    if (MatchesSupersetOf(conflictingField, other))
                     {
                         return;
                     }
                 }
 
                 conflictingFields.Add(other);
+
+                // Determines whether the first field info matches everything that the second field info matches,
+                // in which case the second field info is redundant and doesn't need to be added to the list.
+                static bool MatchesSupersetOf(EnumFieldInfo field1, EnumFieldInfo field2)
```

Comment:
I'll admit I find the use of the term "superset" to be somewhat confusing in this context. I would recommend renaming to something like `MatchesWith` or `ConflictsWith`. Perhaps also rename `field1` and `field2` to `current` and `other` to better emphasize what field is stored and which one is being added.